backend/src/org/commcare/suite/model/OfflineUserRestore.java

```diff
@@ -0,0 +1,182 @@
+            XmlPullParserException, InvalidReferenceException {
+
+        this.reference = reference;
+        checkThatRestoreIsValid();
+        this.password = PropertyUtils.genUUID();
+    }
+
+    public static OfflineUserRestore buildInMemoryUserRestore(InputStream restoreStream)
+            throws UnfullfilledRequirementsException, IOException, InvalidStructureException,
+            XmlPullParserException {
+
+        OfflineUserRestore offlineUserRestore = new OfflineUserRestore();
+        byte[] restoreBytes = StreamsUtil.inputStreamToByteArray(restoreStream);
+        offlineUserRestore.restore = new String(restoreBytes);
+        offlineUserRestore.checkThatRestoreIsValid();
+        offlineUserRestore.password = PropertyUtils.genUUID();
+
+        return offlineUserRestore;
+    }
+
+    public InputStream getRestoreStream() {
+        if (reference != null) {
+            // user restore xml was installed to a file
+            return getStreamFromReference();
+        } else {
+            // user restore xml was installed in memory (CLI)
+            return getInMemoryStream();
+        }
+    }
+
+    private InputStream getInMemoryStream() {
+        try {
+            return new ByteArrayInputStream(restore.getBytes("UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private InputStream getStreamFromReference() {
+        try {
+            Reference local = ReferenceManager._().DeriveReference(reference);
+            return local.getStream();
+        } catch (IOException | InvalidReferenceException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public String getReference() {
+        return reference;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    @Override
+    public void readExternal(DataInputStream in, PrototypeFactory pf)
+            throws IOException, DeserializationException {
+        this.recordId = ExtUtil.readInt(in);
+        this.reference = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        this.restore = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        this.username = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        this.password = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+    }
+
+    @Override
+    public void writeExternal(DataOutputStream out) throws IOException {
+        ExtUtil.writeNumeric(out, recordId);
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(reference));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(restore));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(username));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(password));
+    }
+
+    @Override
+    public void setID(int ID) {
+        recordId = ID;
+    }
+
+    @Override
+    public int getID() {
+        return recordId;
+    }
+
+    private void checkThatRestoreIsValid()
```

Comment:
There's a potential argument to be made that we should actually be including the full restore semantics that we'll use in the app but in "read only" mode somehow. For now that's almost certainly overkill, though.